### PR TITLE
Implement an internal renewal mechanism for token refresh operations

### PR DESCRIPTION
**Pull request by meerd**
*Description of changes:*
Refreshing tokens through webserver reloads can lead to performance bottlenecks and other issues. This new approach aims to move the refresh logic into the enclave, enabling more efficient and optimized token refreshing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by clupuishere — DISMISSED** [review]

**Review by mariusknaust — DISMISSED** [review]

File: src/vtok_p11/src/backend/token.rs [hunk 1]
```diff
@@ -106,6 +106,15 @@ impl Token {
         util::time::monotonic_secs() > self.expiry_ts
     }
 
+    pub fn try_refresh(&mut self, new_expiry_ts: u64) {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_p11/src/backend/token.rs:109]: Will be addressed in the next revision.

File: src/vtok_common/src/config.rs [hunk 2]
```diff
@@ -77,6 +79,14 @@ impl Config {
         Ok(())
     }
 
+    /// Get the modification time of the config file
+    pub fn modif_time() -> Result<SystemTime, Error> {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_common/src/config.rs:83]: Renaming to `modification_time`.

[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_common/src/config.rs:83]: Renaming to `modification_time`.

File: src/vtok_p11/src/backend/token.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_p11/src/backend/token.rs:109]: Addressed in the next revision.

File: src/vtok_p11/src/backend/device.rs [hunk 3]
```diff
@@ -120,16 +125,77 @@ impl Device {
...
+        self.try_reload_slots();
+
         match self.token(slot_id) {
             Ok(token) => token.session(handle),
             Err(_) => None,
         }
     }
 
+    fn reload_slots(&mut self) -> bool {
+        match Config::load_ro().map_err(|_| Error::GeneralError) {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_p11/src/backend/device.rs:144]: This must be a leftover or c/p error. :)

File: src/vtok_p11/src/backend/device.rs [hunk 4]
```diff
@@ -120,16 +125,77 @@ impl Device {
...
+                            }
+                        }
+                        (_, _) => {
+                            // The slot is up-to-date.
+                        }
+                    }
+                }
+                true
+            }
+            Err(_) => {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_p11/src/backend/device.rs:180]: Good point!

File: src/vtok_p11/src/backend/device.rs [hunk 5]
```diff
@@ -120,16 +125,77 @@ impl Device {
...
+                error!("Failed to reload slots: Unable to load configuration.");
+                false
+            }
+        }
+    }
+
+    fn try_reload_slots(&mut self) {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_p11/src/backend/device.rs:187]: Addressed in the next revision.

File: src/vtok_agent/src/agent/mngtok.rs [hunk 6]
```diff
@@ -304,6 +304,62 @@ impl ManagedToken {
...
+                self.label.clone(),
+                self.pin.clone(),
+                self.to_schema_token()?,
+            )
+            .map_err(Error::EnclaveError)?
+            .map_err(Error::UpdateTokenError)?;
+
+        if let Some(_target) = self.target.as_ref() {
+            debug!("Updating managed service configuration");
+            self.satisfy_target(false).or_else(|e| {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_agent/src/agent/mngtok.rs]: Keeping it as is for now.

File: src/vtok_agent/src/agent/mngtok.rs [hunk 7]
```diff
@@ -481,7 +481,12 @@ impl ManagedToken {
                         httpd::HttpdService::write_tls_entries(
                             &path, uid, gid, &key_uri, cert_path,
                         )?;
-                        None
+
+                        if restart_hint {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_agent/src/agent/mngtok.rs:485]: error[E0658]: use of unstable library feature 'bool_to_option'

File: src/vtok_agent/src/agent/mngtok.rs [hunk 8]
```diff
@@ -477,20 +470,17 @@ impl ManagedToken {
                             &path, uid, gid, &key_uri, cert_path,
                         )?;
 
-                        let post_sync_action = if restart_hint {
-                            PostSyncAction::RestartNginx
+                        if restart_hint {
```
[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_agent/src/agent/mngtok.rs:474]: error[E0658]: use of unstable library feature 'bool_to_option'

[the remark this answers is not in the document]
Comment (mariusknaust) [reply src/vtok_agent/src/agent/mngtok.rs:474]: This feature is stable for a while already. Alternative: `restart_hint..then(|| PostSyncAction::RestartNginx)`.

[the remark this answers is not in the document]
Comment (meerd) [reply src/vtok_agent/src/agent/mngtok.rs:474]: `rustc v1.68.2` is complaining about it.

**Review by mariusknaust — APPROVED** [review]

**Review by clupuishere — APPROVED** [review]

**Pull request merged**